File: ios/Porchivo/Services/PendingActionStore.swift
import Foundation

/**
 * UserDefaults-backed queue for ``PendingAction``s.
 * Serialises the entire array as a single `Data` blob so it survives app restarts.
 */
final class PendingActionStore {

    private let defaults = UserDefaults.standard
    private let key = "porchivo_pending_actions"
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    init() {
        encoder.dateEncodingStrategy = .iso8601
        decoder.dateDecodingStrategy = .iso8601
    }

    /// Load all queued actions from disk.
    func loadActions() -> [PendingAction] {
        guard let data = defaults.data(forKey: key) else { return [] }
        return (try? decoder.decode([PendingAction].self, from: data)) ?? []
    }

    /// Persist the full action list to disk.
    func saveActions(_ actions: [PendingAction]) {
        if let data = try? encoder.encode(actions) {
            defaults.set(data, forKey: key)
        }
    }

    /// Remove all queued actions (called on sign-out).
    func clear() {
        defaults.removeObject(forKey: key)
    }
}

File: android/app/src/main/java/com/rork/porchivo/data/PendingActionStore.kt
package com.rork.porchivo.data

import android.content.Context
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/**
 * SharedPreferences-backed queue for [PendingAction]s.
 * Serialises the entire list as a single JSON string so it survives app restarts.
 */
class PendingActionStore(context: Context) {

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true }
    private val serializer = ListSerializer(PendingAction.serializer())

    /** Load all queued actions from disk. */
    fun loadActions(): List<PendingAction> {
        val stored = prefs.getString(KEY, null) ?: return emptyList()
        return try {
            json.decodeFromString(serializer, stored)
        } catch (e: Exception) {
            emptyList()
        }
    }

    /** Persist the full action list to disk. */
    fun saveActions(actions: List<PendingAction>) {
        val encoded = json.encodeToString(serializer, actions)
        prefs.edit().putString(KEY, encoded).apply()
    }

    /** Remove all queued actions (called on sign-out). */
    fun clear() {
        prefs.edit().remove(KEY).apply()
    }

    companion object {
        private const val PREFS_NAME = "porchivo_pending_actions"
        private const val KEY = "pending_actions"
    }
}
